aws-lc-rs/src/rsa/key.rs

```diff
@@ -0,0 +1,490 @@
+    /// `signature`'s length must be exactly the length returned by
+    /// `public_modulus_len()`.
+    ///
+    /// Many other crypto libraries have signing functions that takes a
+    /// precomputed digest as input, instead of the message to digest. This
+    /// function does *not* take a precomputed digest; instead, `sign`
+    /// calculates the digest itself.
+    ///
+    /// # *ring* Compatibility
+    /// Our implementation ignores the `SecureRandom` parameter.
+    ///
+    // # FIPS
+    // The following conditions must be met:
+    // * RSA Key Sizes: 2048, 3072, 4096
+    // * Digest Algorithms: SHA256, SHA384, SHA512
+    //
+    /// # Errors
+    /// `error::Unspecified` on error.
+    /// With "fips" feature enabled, errors if digest length is greater than `u32::MAX`.
+    pub fn sign(
+        &self,
+        padding_alg: &'static dyn RsaEncoding,
+        _rng: &dyn rand::SecureRandom,
+        msg: &[u8],
+        signature: &mut [u8],
+    ) -> Result<(), Unspecified> {
+        let encoding = padding_alg.encoding();
+
+        let mut md_ctx = digest::digest_ctx::DigestContext::new_uninit();
+        let mut pctx = null_mut::<EVP_PKEY_CTX>();
+        let digest = digest::match_digest_type(&encoding.digest_algorithm().id);
+
+        if 1 != unsafe {
+            EVP_DigestSignInit(
+                md_ctx.as_mut_ptr(),
+                &mut pctx,
+                *digest,
+                null_mut(),
+                *self.evp_pkey,
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        if let RsaPadding::RSA_PKCS1_PSS_PADDING = encoding.padding() {
+            // AWS-LC owns pctx, check for null and then immediately detach so we don't drop it.
+            let pctx = DetachableLcPtr::new(pctx)?.detach();
+            super::signature::configure_rsa_pkcs1_pss_padding(pctx)?;
+        }
+
+        let max_len = super::signature::get_signature_length(&mut md_ctx)?;
+
+        debug_assert!(signature.len() >= max_len);
+
+        let computed_signature = compute_rsa_signature(&mut md_ctx, msg, signature)?;
+
+        debug_assert!(computed_signature.len() >= signature.len());
+
+        Ok(())
+    }
+
+    /// Returns the length in bytes of the key pair's public modulus.
+    ///
+    /// A signature has the same length as the public modulus.
+    #[must_use]
+    pub fn public_modulus_len(&self) -> usize {
+        // This was already validated to be an RSA key so this can't fail
+        match self.evp_pkey.get_rsa() {
+            Ok(rsa) => {
+                // https://github.com/awslabs/aws-lc/blob/main/include/openssl/rsa.h#L99
+                unsafe { (RSA_size(*rsa)) as usize }
+            }
+            Err(_) => verify_unreachable!(),
+        }
+    }
+}
+
+impl Debug for KeyPair {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        f.write_str(&format!(
+            "KeyPair {{ public_key: {:?} }}",
```

Comment:
I think it would be better for the debug message to indicate that it's an "RsaKeyPair".